gobblin-modules/gobblin-http/src/main/java/gobblin/http/HttpJoinConverter.java

```diff
@@ -0,0 +1,108 @@
+package gobblin.http;
+
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import lombok.extern.slf4j.Slf4j;
+
+import gobblin.async.AsyncRequest;
+import gobblin.async.AsyncRequestBuilder;
+import gobblin.async.BufferedRecord;
+import gobblin.configuration.WorkUnitState;
+import gobblin.converter.Converter;
+import gobblin.converter.DataConversionException;
+import gobblin.converter.SchemaConversionException;
+import gobblin.converter.SingleRecordIterable;
+import gobblin.writer.WriteCallback;
+
+/**
+ * This converter converts an input record (DI) to an output record (DO) which
+ * contains original input data and http response.
+ *
+ * Sequence:
+ * Convert DI to HttpOperation
+ * Convert HttpOperation to RQ (by internal ConverterRequestBuilder)
+ * Execute http request, get response RP (by HttpClient)
+ * Combine info (DI, RQ, RP, status, etc..) to generate output DO
+ */
+@Slf4j
+public abstract class HttpJoinConverter<SI, SO, DI, DO, RQ, RP> extends Converter<SI, SO, DI, DO> {
+  public static final String CONF_PREFIX = "gobblin.converter.http.";
+
+  protected HttpClient<RQ, RP> httpClient = null;
+  protected ResponseHandler<RP> responseHandler = null;
+  protected AsyncRequestBuilder<HttpOperation, RQ> requestBuilder = null;
+
+  @Override
+  public final SO convertSchema(SI inputSchema, WorkUnitState workUnit)
+      throws SchemaConversionException {
+    httpClient = createHttpClient(workUnit);
+    responseHandler = createResponseHandler(workUnit);
+    requestBuilder = createRequestBuilder(workUnit);
+    return convertSchemaImpl(inputSchema, workUnit);
+  }
+
+  protected abstract HttpClient<RQ, RP>   createHttpClient(WorkUnitState workUnit) throws IllegalStateException;
```

Comment:
On creating these constructs, there is no point to proceed if the state is illegal. The implementations should throw RuntimeException. Remove the `throws`?